test/libsolidity/semanticTests/abiEncoderV2/dirty_memory_dynamic_array_and_calldata_string.sol

```diff
@@ -0,0 +1,34 @@
+pragma abicoder v2;
+
+contract C {
+    struct Data {
+        uint256[] a;
+        string b;
+    }
+
+    function f(Data calldata data, uint256[] calldata a) public pure returns (bytes memory) {
+        // Make memory dirty (numbers come from trial & err)
+        for (uint pos = 0x80; pos < 0x880; pos += 0x20)
+            assembly {
+                mstore(pos, not(0))
+            }
+
+        return abi.encode(data, a);
+    }
+
+    function g(Data memory data, uint256[] calldata a) public pure returns (bytes memory) {
+        // Make memory dirty for string (0x126 is where the string ends)
+        for (uint pos = 0x126; pos < 0x880; pos += 0x20)
```

Comment:
Isn't this function redundant? Looks like the other one makes everything dirty and this one only after the string. But we're only interested in what happens after the string, so both will break if cleanup is broken and both will work if cleanup works.

So unless I'm missing something, I'd be for removing `g()` from this test, especially because it requires more specific hard-coding (you depend on the exact location of the string here). If you do decide to keep this, I think you should at least add an assert that positions 120-126 contain `"123456"`. This way it won't break silently.

Reply:
`g()` uses a memory tuple/struct so this one is testing the copying from memory to memory (where as the other one is for the calldata -> memory variant) 

The "dirty-making"  is so precise here because (I suspect) the memory is already filled before we want to make it dirty with the `data` parameter. When I used different numbers it just gave me a FAILED return..
I am actually not super confident that the dirty-ing is working correct here. If I start at 0x120 I get:

```
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,20]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,60]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,40]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,40]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,80]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,ff]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6]
  [ff,ff,ff,ff,ff,ff,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0] // the ff's here should be the 123456 array.
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,af]
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,bf]

```

Reply:
Ah, right. Missed the fact that it's using a `memory` arg.

> The "dirty-making" is so precise here because (I suspect) the memory is already filled before we want to make it dirty with the data parameter.

Hmm... in that case maybe it would be better to make it a local `memory` variable initialized in the function? Then you could do the dirtying before the variable declaration and be less precise.

In any case an `assert` would be a good validaion that you found the right spot in memory.